nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,253 @@
+  : Node()
+{
+}
+
+nest::EpropArchivingNode::EpropArchivingNode( const EpropArchivingNode& n )
+  : Node( n )
+{
+}
+
+void
+nest::EpropArchivingNode::init_update_history()
+{
+  // register first entry for every synapse, increase access counter if entry already in list
+
+  long shift = get_shift();
+
+  std::vector< HistEntryEpropUpdate >::iterator it =
+    std::lower_bound( update_history_.begin(), update_history_.end(), shift );
+
+  if ( it == update_history_.end() or it->t_ != shift )
+    update_history_.insert( it, HistEntryEpropUpdate( shift, 1 ) );
+  else
+    ++it->access_counter_;
+}
+
+void
+nest::EpropArchivingNode::write_update_to_history( long t_previous_update, long t_current_update )
+{
+  long shift = get_shift();
+
+  std::vector< HistEntryEpropUpdate >::iterator it;
+
+  it = std::lower_bound( update_history_.begin(), update_history_.end(), t_current_update + shift );
+
+  if ( it != update_history_.end() or it->t_ == ( t_current_update + shift ) )
+    ++it->access_counter_;
+  else
+    update_history_.insert( it, HistEntryEpropUpdate( t_current_update + shift, 1 ) );
+
+  it = std::lower_bound( update_history_.begin(), update_history_.end(), t_previous_update + shift );
+
+  if ( it != update_history_.end() or it->t_ == ( t_previous_update + shift ) )
+    --it->access_counter_;
+}
+
+void
+nest::EpropArchivingNode::write_eprop_parameter_to_map( std::string parameter_name, double parameter_value )
+{
+  eprop_parameter_map_[ parameter_name ] = parameter_value;
+}
+
+void
+nest::EpropArchivingNode::write_surrogate_gradient_to_history( long time_step, double surrogate_gradient )
+{
+  eprop_history_.push_back( HistEntryEpropArchive( time_step, surrogate_gradient, 0.0 ) );
+}
+
+void
+nest::EpropArchivingNode::write_error_signal_to_history( long time_step, double error_signal )
+{
+  long shift = delay_out_norm_;
+  eprop_history_.push_back( HistEntryEpropArchive( time_step - shift, 0.0, error_signal ) );
+}
+
+void
+nest::EpropArchivingNode::write_learning_signal_to_history( long& time_step,
+  long& delay_out_rec,
+  double& weight,
+  double& error_signal )
+{
+  long shift = delay_rec_out_ + delay_out_norm_ + delay_out_rec;
+
+  std::deque< HistEntryEpropArchive >::iterator it_hist;
+  std::deque< HistEntryEpropArchive >::iterator it_hist_end;
+
+  get_eprop_history( time_step - shift, &it_hist );
+  get_eprop_history( time_step - shift + delay_out_rec, &it_hist_end );
+
+  if ( it_hist != it_hist_end )
+  {
+    it_hist->learning_signal_ += weight * error_signal;
+    ++it_hist;
+  }
+}
+
+void
+nest::EpropArchivingNode::write_firing_rate_reg_to_history( long t_current_update, double f_target, double c_reg )
+{
+  long const update_interval = kernel().simulation_manager.get_eprop_update_interval().get_steps();
+  long shift = Time::get_resolution().get_steps();
+
+  double f_av = n_spikes_ / static_cast< double >( update_interval );
+  double f_target_ = f_target / 1000.0; // convert to kHz
```

Comment:
As far as we can see, parameters that express rates or frequencies are passed in Hz in NEST models. However, since the update_interval is given in ms, `f_av` is in kHz - thus, we also have to convert f_target to kHz. See, https://github.com/nest/nest-simulator/pull/2867/commits/abf5eb361e9e0b65914492034d2cdfea6e78b878 for an updated and more clear implementation of this function.

Reply:
Ok, I understand your argument. But I have more questions:

1. You get `update_interval` in steps. But then f_av is not in kHz, but in "spikes per step", whatever the step length might be. Shouldn't you get the update interval in ms and divide by that?
2. Below, you divide a difference of rates by the interval length, so you get essential "spikes per time^2". Is that intended?

Reply:
That's correct, please see the updated version that I linked in my answer, where we do exactly that.

Reply:
Now I am getting confused by Github—does it not show the newest code here? In your new code, you now have `.../update_interval * dt = ... / (update_interval / dt)`, where `update_interval` is in ms, so by dividing by `dt` you  are again back to steps, so `f_av` is still spikes per step.

Reply:
Github apparently does not show the newest code here. That is true, there is something not right - I need to think again about this.

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/2867/commits/b5ab4b4205371188e2048ce4369a27e20ff071e4.